src/runtime-rs/crates/hypervisor/src/dragonball/inner.rs

```diff
@@ -0,0 +1,300 @@
+
+        // add pending devices
+        while let Some(dev) = self.pending_devices.pop() {
+            self.add_device(dev).await.context("add_device")?;
+        }
+
+        // start vmm and wait ready
+        self.start_vmm_instance().context("start vmm instance")?;
+        self.wait_vmm_ready(timeout).context("wait vmm")?;
+
+        Ok(())
+    }
+
+    pub(crate) fn run_vmm_server(&mut self, _timeout: i32) -> Result<()> {
+        if !self.config.jailer_path.is_empty() {
+            self.jailed = true;
+        }
+
+        // create jailer root
+        create_dir_all(self.jailer_root.as_str())
+            .map_err(|e| anyhow!("Failed to create dir {} err : {:?}", self.jailer_root, e))?;
+
+        // create run dir
+        self.run_dir = [RUN_PATH_PREFIX, self.id.as_str()].join("/");
+        create_dir_all(self.run_dir.as_str())
+            .with_context(|| format!("failed to create dir {}", self.run_dir.as_str()))?;
+
+        // run vmm server
+        self.vmm_instance
+            .run_vmm_server(&self.id, self.netns.clone())
+            .context("run vmm server")?;
+        self.state = VmmState::VmmServerReady;
+
+        Ok(())
+    }
+
+    pub(crate) fn cleanup_resource(&self) {
+        if self.jailed {
+            self.umount_jail_resource(DRAGONBALL_KERNEL).ok();
+            self.umount_jail_resource(DRAGONBALL_ROOT_FS).ok();
+            for id in &self.cached_block_devices {
+                self.umount_jail_resource(id.as_str()).ok();
+            }
+        }
+
+        std::fs::remove_dir_all(&self.vm_path)
+            .map_err(|err| {
+                error!(sl!(), "failed to remove dir all for {}", &self.vm_path);
+                err
+            })
+            .ok();
+    }
+
+    fn set_vm_base_config(&mut self) -> Result<()> {
+        let serial_path = [&self.run_dir, "console.sock"].join("/");
+        let vm_config = VmConfigInfo {
+            serial_path: Some(serial_path),
+            mem_size_mib: self.config.memory_info.default_memory as usize,
+            vcpu_count: self.config.cpu_info.default_vcpus as u8,
+            ..Default::default()
+        };
+        info!(sl!(), "vm config: {:?}", vm_config);
+
+        self.vmm_instance
+            .set_vm_configuration(vm_config)
+            .context("set vm configuration")
+    }
+
+    pub(crate) fn umount_jail_resource(&self, jailed_path: &str) -> Result<()> {
+        let path = [self.jailer_root.as_str(), jailed_path].join("/");
+        nix::mount::umount2(path.as_str(), nix::mount::MntFlags::MNT_DETACH)
+            .with_context(|| format!("umount path {}", &path))
+    }
+
+    pub(crate) fn get_resource(&self, src: &str, dst: &str) -> Result<String> {
+        if self.jailed {
+            self.jail_resource(src, dst)
+        } else {
+            Ok(src.to_string())
+        }
+    }
+
+    fn jail_resource(&self, src: &str, dst: &str) -> Result<String> {
+        info!(sl!(), "src {} dst {}", src, dst);
```

Comment:
Nit: Suggest adding `jail resource in the log.

Reply:
Thanks for your review, already updated.